### PR TITLE
New /api/related/{uuid} Endpoint

### PR DESCRIPTION
**Pull request by x65han**
Screen Shot 2020-04-29 at 3 46 27 PM

Example: http://127.0.0.1:8000/api/related/bf2jg3qy?pageNumber=2

### CODE REVIEW

File: api/requirements.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ httpx==0.12.1
 pyserini==0.8.1
 python-dotenv==0.12.0
 spacy==2.2.4
-tensorflow-gpu==2.1.0
 torch==1.4.0
 transformers==2.7.0
+hnswlib==0.3.4
```
Comment (edwinzhng) [inline api/requirements.txt:9]: can you keep these in alphabetical order

File: api/app/models.py [hunk 2]
```diff
@@ -27,21 +28,41 @@ def validate_highlights(cls, v, values):
...
     cord19 = 'cord19'
     trialstreamer = 'trialstreamer'
+
+
+class RelatedArticle(BaseModel):
+    distance: str
+    uid: str
+    url: str
+    title: str
+    authors: str = None
```
Comment (edwinzhng) [inline api/app/models.py:60]: should be `List[str] = []`

File: api/app/services/relatedHelper.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
+from app.settings import settings
+import hnswlib
+import csv
+
+
+class RelatedHelper:
```
Comment (edwinzhng) [inline api/app/services/relatedHelper.py:6]: maybe can should name this `NearestNeighbourSearcher` or `RelatedSearcher`?

File: api/app/routers/related.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            'journal': gen_metadata_from_uid(uid, 'journal'),
+            'publish_time': gen_metadata_from_uid(uid, 'publish_time'),
+            'abstract': gen_metadata_from_uid(uid, 'abstract'),
+        })
+
+    # Generate UUID for query.
+    query_id = str(uuid4())
+
+    # Log query and results.
+    forwarded_header = 'X-Forwarded-For'
```
Comment (edwinzhng) [inline api/app/routers/related.py:48]: can you pull this out into a function maybe in a `util.py` like `get_request_ip` and reuse in the `search.py` file as well?

File: api/app/routers/related.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    request_ip = request.client.host
+    if forwarded_header in request.headers:
+        request_ip = request.headers[forwarded_header]
+
+    related_logger.info(json.dumps({
+        'query_id': query_id,
+        'uid': uid,
+        'pageNumber': pageNumber,
```
Comment (edwinzhng) [inline api/app/routers/related.py:56]: `page_number` to be consistent

File: api/app/routers/related.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+from app.services.relatedHelper import relatedHelper
+from datetime import datetime
+
+router = APIRouter()
+related_logger = build_timed_logger(
+    'related_logger', settings.related_log_path)
+
+
+@router.get('/related/{uid}', response_model=RelatedQueryResponse)
+async def get_related(request: Request, uid: str, pageNumber: int = 1):
```
Comment (edwinzhng) [inline api/app/routers/related.py:16]: `pageNumber` -> `page_number`

File: api/app/services/relatedHelper.py (hunk 3 above)
Comment (x65han) [reply api/app/services/relatedHelper.py:6]: `RelatedSearcher` for now.

File: api/app/routers/related.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    return None
+
+
+def get_authors_from_uid(uid) -> List[str]:
+    authors = gen_metadata_from_uid(uid, 'authors')
+    if authors is None:
+        return []
+
+    return authors.split(';')
```
Comment (edwinzhng) [inline api/app/routers/related.py:77]: strip all whitespace too `[author.strip() for author in authors.split(';')]`

**Pull request merged**